File: arrays/04-Product-of-array-except-self/Solution.java
package com.sandbox;

public class Solution {
  public int[] productExceptSelf(final int[] numbers) {
    assert (numbers.length > 1);
    int current = 0;
    int max = numbers.length - 1;
    int[] result = new int[numbers.length];
    Integer sum = null;
    
    while (true) {
      for (int i = 0; i <= max; i++) {
        if (i == current) continue;
        if (sum == null) {
          sum = numbers[i];
        } else {
          sum *= numbers[i];
        }
      }
      
      result[current] = sum;
      current++;
      sum = null;
      
      if (current == numbers.length) {
        break;
      }
    }
    
    return result;
  }
}

File: arrays/04-Product-of-array-except-self/solution.go
package main

func productExcept(nums []int, idx int) int {
	sum := 1

	for i := 0; i < len(nums); i++ {
		if i == idx {
			continue
		}

		sum *= nums[i]
	}

	return sum
}

// TODO: optimizie to run in O(n) time
func ProductOfArrayExcept(nums []int) []int {
	max := len(nums)
	result := []int{}

	for i := 0; i < max; i++ {
		result = append(result, productExcept(nums, i))
	}

	return result
}
